### PR TITLE
SPEC-857: retryWrites is now a MongoClient option

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/SPEC-857

### CODE REVIEW

File: source/retryable-writes/retryable-writes.rst [hunk 1]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
-
-SessionOptions
-   Data structure defined in the Driver Session specification to provide options
-   when creating a DriverSession. This could be a strongly typed class in some
-   drivers, or it could be a loosely typed dictionary in other drivers.
+   Driver object representing a client session, which is defined in the Driver
+   Session specification. This object is always associated with a server
+   session; however, drivers will pool server sessions so that creating a
+   DriverSession will not always entail creation of a new server session. The
+   name of this object MAY vary across drivers.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Removed SessionOptions entirely and referenced Driver Session spec re: creation of DriverSession objects. The bit about pooling is relevant to our use case below where we might create these internally for a single operation.

File: source/retryable-writes/retryable-writes.rst [hunk 2]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
 -----------------
 
-This specification defines the name for a new DriverSession option,
-``retryWrites``. To the extent possible, drivers SHOULD use the defined name.
-However, where a driver’s or language’s naming conventions would conflict,
-drivers SHOULD honor their existing conventions. For example, a driver may use
-``retry_writes`` instead of ``retryWrites``.
+This specification defines the name for a new MongoClient option,
+``retryWrites``. Drivers MUST use the defined name to ensure portability of
+connection strings across applications and drivers.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: I didn't find "Naming Deviations" discussed in other specs that talk about URI options, but I think we don't want flexibility here in order to keep connection strings portable.

File: source/retryable-writes/retryable-writes.rst [hunk 3]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
 
-Configuration
--------------
+MongoClient Configuration
+-------------------------
 
-The following option MUST be added to the SessionOptions data structure:
+This specification introduces the following client-level configuration option.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Other specs (e.g. Server Selection) don't mention "connection string" by name, so I kept the same format here.

File: source/retryable-writes/retryable-writes.rst [hunk 4]
```diff
@@ -120,13 +111,22 @@ Behavioral Changes for Write Commands
...
-method MUST automatically add a transaction ID when executed within a
-DriverSession where retryable writes have been enabled.
+`Read and Write Concern`_ specifications) MUST also accept an optional
+DriverSession parameter. If a DriverSession parameter is specified by the
+application, drivers MUST use it to generate the transaction ID for a retryable
+write operation. Otherwise, drivers MUST internally construct a new
+DriverSession for the sole purpose of generating a transaction ID. Any
+internally constructed DriverSession SHOULD be destroyed as soon as the
+operation is complete in its interactions with the server so that the
+DriverSession may return its server session to the pool.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: "complete in its interactions with the server" seemed like a portable way to address all drivers, since the helper method returning might not apply in the async case. If a write operation is complete in its interactions with the server, that means it either succeeded (after possible retry attempts) or we've hit an error that we're throwing back to the user.

File: source/retryable-writes/retryable-writes.rst [hunk 5]
```diff
@@ -274,8 +275,8 @@ This includes an `update`_ command where any statement in the updates sequence
 specifies a ``multi`` option of ``true`` or a `delete`_ command where any
 statement in the ``deletes`` sequence specifies a ``limit`` option of ``0``. In
 the context of the `CRUD`_ specification, this includes the ``updateMany()`` and
-``deleteMany()`` methods. Drivers MUST rely on the server to raise an error if
-multi-document write operations are not supported. 
+``deleteMany()`` methods. Drivers MUST rely on the server to raise an error if a
+multi-document write operations would be retried and is not supported.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: I have a TODO item below in the spec where I'd like to discuss exactly how we handle these unsupported write operations. Please use that comment thread for discussion.

File: source/retryable-writes/retryable-writes.rst [hunk 6]
```diff
@@ -415,13 +417,13 @@ The design of this specification piggy-backs that of the Driver Session
...
 
 Drivers will rely on the server to yield an error if an unsupported write
-operation is executed within a DriverSession where retryable writes have been
-enabled. This will free drivers from having to maintain a list of supported
-write operations and also allow for forward compatibility when future server
-versions begin to support retryable behavior for additional write operations.
+operation would be retried and is not supported. This will free drivers from
+having to maintain a list of supported write operations and also allow for
+forward compatibility when future server versions begin to support retryable
+behavior for additional write operations.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: This may certainly change based on the TODO comment thread below.

File: source/retryable-writes/retryable-writes.rst [hunk 7]
```diff
@@ -511,6 +513,15 @@ experience across all drivers.
...
 
+.. todo:: Determine how this should change. Do we end up using whitelists for
+   supported operations (based on wire protocol verison) and avoid injecting
+   transaction IDs on all write ops, or rely on the server to raise an error if
+   unsupported ops would be retried? In both cases, it's possible for users to
+   be unaware that some operations are not supported until some error in prod.
+   Given the MongoClient-wide scope of retryWrites, we likely don't want to
+   prohibit unsupported operations outright (either in the client API or by
+   having the server raise an exception on the first attempt).
+
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Let's discuss how we'd like to handle unsupported operations. This was an outstanding question following the design meetings earlier in the week.

Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: I guess we have to whitelist supported operations and only include txnNum if the command is in the whitelist. Document carefully in the MongoDB Manual which operations are retryable in which server versions. The alternative, which is to require that users only execute retryable operations or else receive an error, might make retryable writes too hard to use in applications with a lot of existing code. For future work we can specify a format in the isMaster reply that communicates the server's list of retryable write operations to the driver, which will relieve us of the burden, but for 3.6 we can hardcode the whitelist in all drivers.

If we follow this approach, I propose we allow unacknowledged writes with a retryWrites client. They're not retryable, obviously, but we should allow them for consistency with other non-retryable operations.

I am very ambivalent about this opinion, however.

File: source/retryable-writes/retryable-writes.rst [hunk 8]
```diff
@@ -31,9 +31,8 @@ write operation(s) and allow for retrying the operation if the driver fails to
 obtain a command result (e.g. network error after a replica set failover). This
 specification will outline how an API for retryable write operations will be
 implemented in drivers. The specification will define an option to enable
-retryable writes when constructing a client session and describe how a
-transaction ID will be provided to write commands executed within that
-client session.
+retryable writes for an application and describe how a transaction ID will be
+provided to write commands executed therein.
```
Comment (rstam) [inline source/retryable-writes/retryable-writes.rst]: Apparently you can only comment on lines that are changed... that sucks.

In the paragraph before this one (line 24)  and throughout this document replace DriverSession (and "driver session" etc..) with ClientSession.

The sessionId is no longer opaque (line 23). Drivers must know the format because they can now be generated by the driver.

File: source/retryable-writes/retryable-writes.rst [hunk 9]
```diff
@@ -137,21 +137,20 @@ check the user’s command document to determine if it is a write, nor check
 whether the server is new enough to support a transaction ID for the command.
 The method should simply send the user’s command document to the server as-is.
 
-This specification does not affect write commands executed outside of a
-DriverSession or within a DriverSession where retryable writes have not been
-enabled.
+This specification does not affect write commands executed in an application
+where retryable writes have not been enabled.
```
Comment (rstam) [inline source/retryable-writes/retryable-writes.rst]: (This comment is for lines 134-138, but github won't let me add it there).

This contradicts the goal that from an ops point of view we want every operation (read or write) to have a sessionId.

It should not be the application's responsibility to add the sessionId when using the runCommand helper.

File: source/retryable-writes/retryable-writes.rst [hunk 10]
```diff
@@ -229,6 +228,8 @@ user.
 Supported Write Operations
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+MongoDB 3.6 will support retryability for some, but not all, write operations.
```
Comment (rstam) [inline source/retryable-writes/retryable-writes.rst]: (This comment is for lines 223-226 but github won't let me add it there).

I'm wondering why we don't raise the original error when the retry fails.

File: source/retryable-writes/retryable-writes.rst (hunk 7 above)
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Oh that's a better idea. I agree with @rstam .

File: source/retryable-writes/retryable-writes.rst [hunk 11]
```diff
@@ -137,21 +137,20 @@ check the user’s command document to determine if it is a write, nor check
...
-When constructing any write command that will be executed within a DriverSession
+When constructing any write command that will be executed within a MongoClient
 where retryable writes have been enabled, drivers MUST increment the transaction
 number for the corresponding server session and include the server session ID
 and transaction number in top-level ``sessionId`` and ``txnNum`` fields,
-respectively. The ``sessionId`` value is an opaque BSON type and MUST be
-obtained from the ``startSession`` command (as discussed in the Driver Session
-specification). The ``txnNum`` value MUST be an unsigned integer and the server
-will accept 32-bit (BSON type 0x10) or 64-bit (0x12) values.
+respectively. The ``sessionId`` value is an opaque BSON type (as discussed in
```
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: They had a discussion about it in one of the documents, IIRC. They know that other global command parameters are $-prefixed and have decided that these parameters are not.

File: source/retryable-writes/retryable-writes.rst (hunk 10 above)
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Agreed, we should raise the original error.

File: source/retryable-writes/retryable-writes.rst [hunk 12]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
-    /**
-     * If true, retryable behavior will be applied to write commands
-     * executed within the session.
-     *
-     * Defaults to false, which implies no change in write behavior.
-     */
-    retryWrites: Optional<Boolean>;
-  }
+This boolean option determines whether retryable behavior will be applied to all
+write operations executed within the application. This option MUST default to
```
Comment (behackett) [inline source/retryable-writes/retryable-writes.rst]: Within the context of the client. It's common for applications to create more than one instance of MongoClient.

File: source/retryable-writes/retryable-writes.rst [hunk 13]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
+false, which implies no change in write behavior.
+
+This option MUST NOT be configurable at the level of a database object,
+collection object, or at the level of an individual write operation.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: The take-away from our previous spec review meeting was to only have it as a client-level option. I'd like to be conservative here and avoid having drivers add the option everywhere, even if there is no technical limitation against doing so. A limited introduction also reduces the burden of API changes and documentation.

Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Agreed.

File: source/retryable-writes/retryable-writes.rst [hunk 14]
```diff
@@ -52,65 +51,57 @@ Transaction ID
    The transaction ID identifies the transaction as part of which the command is
    running. In a write command where the client has requested retryable
    behavior, it is expressed by the top-level ``sessionId`` and ``txnNum``
-   fields. The ``sessionId`` component is the corresponding server session ID,
-   which is an opaque BSON value returned by the ``startSession`` command. The
-   ``txnNum`` component is a monotonically increasing (per server session),
+   fields. The ``sessionId`` component is the corresponding server session ID.
+   which is an opaque BSON value defined in the Driver Session specification.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: I'm removing "opaque" language from this spec but would still like to refrain from discussing the `sessionId`'s structure in this spec and instead refer to the Driver Session specification as the authority.

File: source/retryable-writes/retryable-writes.rst (hunk 1 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Limiting `retryWrites` to a global client-level option is intentional. See: https://github.com/mongodb/specifications/pull/174/files#r135094282

File: source/retryable-writes/retryable-writes.rst (hunk 2 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Is MongoClientSettings what your driver uses for a dictionary of options taken by MongoClient in addition to the URI string? I assume those options are then merged with values from the URI query string in some fashion to produce the final options used for the MongoClient?

IIRC, the URI options are case-insensitive (unless that's just libmongoc). PHP happens to use the canonical, camel-case names of the URI options as our options dictionary keys, so I didn't stop to think that naming deviations would come up here. In any event, I'll reword this to allow flexibility for option names outside the URI string.

File: source/retryable-writes/retryable-writes.rst [hunk 15]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
-maintaining a monotonically increasing transaction number for each server
-session used by a DriverSession object created with the ``retryWrites`` option.
-Drivers that pool server sessions MUST preserve the transaction number when
-reusing a server session from the pool with a new DriverSession.
+obtained from the DriverSession object, which will have either been passed to
+the write operation from the application or constructed internally for the
+operation. Drivers will be responsible for maintaining a monotonically
+increasing transaction number for each server session used by a DriverSession
+object. Drivers MUST preserve the transaction number when reusing a server
+session from the pool with a new DriverSession.
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Will do.

File: source/retryable-writes/retryable-writes.rst (hunk 9 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: > If your driver offers a generic command method on your database object, it MUST NOT automatically add a transaction ID. The generic command method MUST NOT check the user’s command document to determine if it is a write, nor check whether the server is new enough to support a transaction ID for the command. The method should simply send the user’s command document to the server as-is.

IIRC, this paragraph was added during the draft stage based on @ajdavis' feedback.

In this section, I'm alluding to read and write commands as indicated in the Read and Write Concern spec (a read command is one that solicits a `readConcern` parameter). To that end, there are commands that are neither read nor write, such as `ping` and `isMaster`. A generic command helper is the only driver facility for sending a raw command to the server, and therefore should not concern itself with injecting options into the command document.

From the context of libmongoc, there are methods for executing arbitrary read, write, and read/write commands. These function like `runCommand()` but _do_ inject read- and write-specific options such as `readConcern`.

I'd argue that the Driver Session specification is free to mandate that `runCommand()` always inject a session ID, but we don't need to mandate injection of a transaction ID (i.e. the combination of session ID and transaction number) in this spec.

File: source/retryable-writes/retryable-writes.rst (hunk 11 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: What other global parameters exist beside `$readPreference`?

File: source/retryable-writes/retryable-writes.rst (hunk 10 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: > When retrying a write command, drivers MUST resend the command with the same transaction ID. Drivers MAY resend the original wire protocol message (see: Can drivers resend the same wire protocol message on retry attempts?). If the second attempt also fails, drivers MUST raise its corresponding error to the user.

In practice, the original error could only be a NetworkException. Since server selection occurs between the first and second attempt and we may have selected a new primary server, what value is there in throwing the original NetworkException? It is likely to reflect the current state of the system. Consider:

 * Node A went offline and the replica set fails over to node B
 * Driver, using outdated topology information, sends a write to A and receives a NetworkException
 * Driver performs server selection and realizes that B is the new primary
 * Node A recovers and is reelected and B goes offline
 * Driver retries the write with B and receives a NetworkException

If B never went offline, I suppose we'd get a server response that it's not a master and simply throw that to the user; however, if it is offline and we get a NetworkException, I think that's more relevant to the user than the previous NetworkException for A. Perhaps this is all a moot point because an election would never complete in the time between server selection and the retry attempt.

Either way, I should clarify that "fails" in "If the second attempt also fails" is only referring to another NetworkException. If the retry attempt receives an `{ ok: 0 }` response from the server indicating that the command failed (e.g. duplicate key exception), we obviously want to throw _that_ to the user instead of the original NetworkException.

File: source/retryable-writes/retryable-writes.rst (hunk 7 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: I concur with the objections to whitelists.

@ajdavis makes a good point that `{ w: 0 }` will need to be permitted in the interest of application portability.

Having the server ignore the transaction ID (sessionID/txnNum pair) on first attempts and erroring on the second attempt still has the caveat of users not realizing that some operations cannot be retried during testing. I know that we settled on only soliciting `retryWrites` as a client-level option (https://github.com/mongodb/specifications/pull/174/files#r135094282), but allowing it on the database and collection objects could help address the "didn't realize until production" situation if that's still a concern. I believe when we last discussed not supporting the option on database and collection objects, it was to avoid creating a dependency on the ThreadLocal driver session object; however, since the drivers now fetch a server session from the pool as needed per operation, ThreadLocal bundling is no longer an issue.

Is this something we'd like to discuss further, or should we just formally request that the server changes its behavior to:

> The server should only return an error if a retry is actually attempted and not supported.

File: source/retryable-writes/retryable-writes.rst (hunk 2 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: Addressed in https://github.com/mongodb/specifications/pull/174/commits/bccedcec582745e1869c97455c78461cb611066b.

File: source/retryable-writes/retryable-writes.rst (hunk 8 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/ed490e62c9c759f74ae5f30fd3546c9f0bed28d3 renames DriverSession to ClientSession.

https://github.com/mongodb/specifications/pull/174/commits/ea6faf0bdb9d7584bbf68b8a8323b858b0d8aefd removes the "opaque" language.

File: source/retryable-writes/retryable-writes.rst (hunk 14 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/ea6faf0bdb9d7584bbf68b8a8323b858b0d8aefd removes the "opaque" language.

File: source/retryable-writes/retryable-writes.rst [hunk 16]
```diff
@@ -52,65 +51,57 @@ Transaction ID
...
-   across drivers.
-
-SessionOptions
-   Data structure defined in the Driver Session specification to provide options
-   when creating a DriverSession. This could be a strongly typed class in some
-   drivers, or it could be a loosely typed dictionary in other drivers.
+   Driver object representing a client session, which is defined in the Driver
+   Session specification. This object is always associated with a server
+   session; however, drivers will pool server sessions so that creating a
+   DriverSession will not always entail creation of a new server session. The
```
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/ed490e62c9c759f74ae5f30fd3546c9f0bed28d3 renames DriverSession to ClientSession.

File: source/retryable-writes/retryable-writes.rst (hunk 15 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/61eb2db2283f5bd91f3af03acbe10334cf18ebfb adds a suggestion to track transaction number on the server session object.

File: source/retryable-writes/retryable-writes.rst (hunk 12 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/7d4b37456c99dbb21a446b5c83ef3a642e742e19 fixes this language.

**Comment by jmikola** [comment]
I commented on each open thread and added a few commits to resolve outstanding change requests. In case you find it difficult to view collapsed comment threads, here's a summary of the open threads:

https://github.com/mongodb/specifications/pull/174#discussion_r134750385 - Conclusion that `retryWrites` is only a client-level option, not configurable at the database, collection, and operation level.

https://github.com/mongodb/specifications/pull/174#discussion_r134001210 - Allowing unacknowledged writes (simply ignore retry option) and requesting that the server change its behavior to ignore transaction IDs for unsupported operations on the first attempt but raise an error on the retry attempt

https://github.com/mongodb/specifications/pull/174#discussion_r134781620 - Generic `runCommand()` helper should not attempt to detect write commands and inject transaction IDs. Similarly, it should not detect read and write commands to inject client- and database-level `readConcern` and `writeConcern` options, respectively.

https://github.com/mongodb/specifications/pull/174#discussion_r134751354 - The server team decided not to $-prefix the session ID and transaction number fields. Offhand, what other global command parameters use $-prefixing other than `$readPreference`?

https://github.com/mongodb/specifications/pull/174#discussion_r134782214 - Discussing whether the driver should throw the first attempt's NetworkException if the retry attempt fails. Either way, we should clarify "failued" for a second attempt only refers to a network error. Any `{ ok: 0 }` response from the server is something we should immediately throw to the user.

File: source/retryable-writes/retryable-writes.rst (hunk 7 above)
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Apparently it is *not* a goal now to help users realized that some operations cannot be retried. We've abandoned that in favor of this goal: make it possible to turn on retryability for a subset of operations, with minimum code changes. Therefore, we must allow unacknowledged writes even when retryWrites is true.

For acknowledged writes that aren't retryable on the server, there's a problem. The server can't *know* whether it's already executed any particular operation or not. E.g., I send an updateMany with a txnNum, and I can't read the server response. Two options:

1. I send it again. The server does not store a transaction record for a non-retryable operation, so it doesn't know whether it's executed it before. Therefore the server may execute the updateMany twice.
2. The driver maintains a whitelist (sorry) of retryable operations and only re-sends operations that are in the whitelist.

File: source/retryable-writes/retryable-writes.rst (hunk 9 above)
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Agreed.

File: source/retryable-writes/retryable-writes.rst (hunk 11 above)
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: $clusterTime. On the other hand maxTimeMS is also global and not $-prefixed.

**Comment by ajdavis** [comment]
Thanks so much for making that list of links, that's a great idea.

File: source/retryable-writes/retryable-writes.rst (hunk 7 above)
Comment (jmikola) [inline source/retryable-writes/retryable-writes.rst]: > This is a major behavior change compared to current drivers: it reduces lost updates and increases double-updates.

This highlights an interesting problem with having the server only return an error on retry attempts. The risk of double updates calls to mind your [blog post](https://emptysqua.re/blog/how-to-write-resilient-mongodb-applications/) and the "may overcount" risk of blindly retrying non-idempotent updates after transient network errors. IMO it's a deal-breaker.

That leaves two options:

 * Command whitelist
 * Ability to configure `retryWrites` at narrower levels than the client (e.g. database, collection, operation) to allow users to disable it when they need to perform an unsupported operation and vice versa

The second bullet still seems like it supports the goal of "make it possible to turn on retryability for a subset of operations, with minimum code changes", but I'm inferring "subset" as operations executed within a configured driver object (of any type) instead of operations of a supported type (executed from anywhere).

Comment (behackett) [reply source/retryable-writes/retryable-writes.rst]: This seems like an easy problem to solve, don't retry unacknowledged writes and document that we don't retry unacknowledged writes and why.

Comment (ajdavis) [reply source/retryable-writes/retryable-writes.rst]: Bernie's comment *does* offer a simple solution for unacknowledged writes.

As for acknowledged writes. "Having the server only return an error on retry attempts" isn't even possible. When the server receives a message from the client for a non-retryable operation, the server doesn't know whether this is the first or second attempt at all. It can't return an error on the second attempt, therefore.

Jeremy I agree that the two options you listed are the two options available to us. Let's just ask Andy and Eliot which they want?

Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: @rstam, @jyemin, @ajdavis, and I spoke with @andy10gen earlier and concluded that drivers should implement a whitelist of commands that support retriability. I will update the spec accordingly and add an explanation of the design decision to the spec as well.

File: source/retryable-writes/retryable-writes.rst [hunk 17]
```diff
@@ -50,11 +50,11 @@ Terms
 Transaction ID
    The transaction ID identifies the transaction as part of which the command is
    running. In a write command where the client has requested retryable
-   behavior, it is expressed by the top-level ``sessionId`` and ``txnNum``
-   fields. The ``sessionId`` component is the corresponding server session ID.
-   which is a BSON value defined in the Driver Session specification. The
-   ``txnNum`` component is a monotonically increasing (per server session),
-   positive 64-bit integer.
+   behavior, it is expressed by the top-level ``lsid`` and ``txnNum`` fields.
+   The ``lsid`` component is the corresponding server session ID. which is a
```
Comment (jyemin) [inline source/retryable-writes/retryable-writes.rst]: remove the period before "which"

File: source/retryable-writes/retryable-writes.rst [hunk 18]
```diff
@@ -345,7 +347,7 @@ Drivers MUST add an optional ``transactionId`` field to the
   transactionId: Document;
```
[the remark this answers is not in the document]
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: Does @rstam require adding a `sessionId` field for APM? If so, that may make half of this value redundant.

I believe the command document only exists in CommandStartedEvent, so users would need to pull out the field from the first event handler to determine if it's a retryable write. I can be convinced to remove this if others feel strongly about it.

File: source/retryable-writes/retryable-writes.rst [hunk 19]
```diff
@@ -533,20 +544,18 @@ Most `CRUD`_ operations are supported apart from ``updateMany()``,
...
+transactions and multi-document writes. MongoDB 3.6 lacks support for retrying
+some CRUD operations, such as ``updateMany()`` and
+``deleteMany()``. Additionally, write commands other than ``insert``,
+``update``, and ``delete`` are not supported at all.
 
-DriverSessions will be used for features other than retryable writes, such as
-causally consistent reads. Consensus among these related specifications is that
-any and all DriverSession options will default to false and users may opt in to
-particular features as desired.
+Furthermore, we cannot know for sure what the server-side overhead will be for
```
[the remark this answers is not in the document]
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/998540740d8808f0f73173f6fd416fb4489718e1 revises this text.

File: source/retryable-writes/retryable-writes.rst (hunk 10 above)
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/bf0a17b573b0a7f97ef00b9b88c274736a056590 clarifies that "fails" refers to a network error

File: source/retryable-writes/retryable-writes.rst (hunk 7 above)
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/8eb5d59b8309c9be85142b5f0e6c2f923d94a85a changes the spec to utilize a whitelist of supported operations

**Comment by jmikola** [comment]
Outstanding questions:

https://github.com/mongodb/specifications/pull/174#discussion_r135340901

Should this spec not suggest that a `transactionId` field be added to APM event classes? The driver session spec doesn't discuss adding a `sessionId` on event classes, and @jyemin noted that the full command document in CommandStartedEvent will have both the `lsid` and `txnNum` fields if users want to access this information.

https://github.com/mongodb/specifications/commit/8eb5d59b8309c9be85142b5f0e6c2f923d94a85a#commitcomment-23880124

@jyemin commented that a paragraph about generating ClientSession objects for a single operation was over-specified. I deleted the paragraph in https://github.com/mongodb/specifications/pull/174/commits/f903438d40bec9c31c14427ef4ce7a2b03c0cac3.

File: source/retryable-writes/retryable-writes.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (ajdavis) [reply source/retryable-writes/retryable-writes.rst]: Let's remove this special transactionId field, and let APM users extract the txnNum from the command document itself. We can add the field later if asked.

[the remark this answers is not in the document]
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/cb3f58086f7d672a2d3993a3a3a6b3f2847f1e07 removes the `transactionId` field on APM event objects

File: source/retryable-writes/retryable-writes.rst [hunk 20]
```diff
@@ -242,8 +237,8 @@ These methods above are defined in the `CRUD`_ specification.
 
 Later versions of MongoDB may add support for additional write operations.
 
-Unsupported Write Operations (client-side error)
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Unsupported Write Operations
```
Comment (behackett) [inline source/retryable-writes/retryable-writes.rst]: This section does a great job of telling you what write operations aren't supported, and therefore which to not add a transaction id to, but it never says not to retry these write operations.

Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/25623b6dcd077edae6704e12d526cbb5c397d6d6 adds clarification.

File: source/retryable-writes/retryable-writes.rst (hunk 10 above)
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: > If the server selected for a retry attempt does not support retryable writes (e.g. mixed-version cluster), retrying is not possible and drivers MUST raise the original network error to the user.

@jyemin: the spec language above and corresponding pseudo-code doesn't address a server selection failure. Based on the current pseudo-code, I expect that `selectServer("writable")` before the retry attempt would throw an exception if it fails to return a writable server. I'd agree that failing to select a writable server for the retry attempt is not much different (from the user's perspective) as a selected server not support retryable writes, so throwing the original network exception works for me.

Comment (jyemin) [reply source/retryable-writes/retryable-writes.rst]: Let's leave it as is.

**Comment by xdg** [comment]
abstain

File: source/retryable-writes/retryable-writes.rst [hunk 21]
```diff
@@ -253,29 +248,23 @@ server may not support retryable writes if the
...
-the collection is instantiated with an unacknowledged write concern from a
-DriverSession where retryable writes have been enabled.
-
-Unsupported Write Operations (server-side error)
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Write commands specifying an unacknowledged write concern (e.g. ``{w: 0})``) do
+not support retryable behavior. Drivers MUST NOT add a transaction ID to any
+write command with an unacknowledged write concern executed within a MongoClient
+where retryable writes have been enabled. Drivers MUST NOT retry these commands
+if they fail to return a response.
```
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Delete the last sentence: these are unacknowledged.

File: source/retryable-writes/retryable-writes.rst [hunk 22]
```diff
@@ -253,29 +248,23 @@ server may not support retryable writes if the
...
 Write commands where a single statement might affect multiple documents will not
 be initially supported by MongoDB 3.6, although this may change in the future.
 This includes an `update`_ command where any statement in the updates sequence
 specifies a ``multi`` option of ``true`` or a `delete`_ command where any
 statement in the ``deletes`` sequence specifies a ``limit`` option of ``0``. In
 the context of the `CRUD`_ specification, this includes the ``updateMany()`` and
-``deleteMany()`` methods. Drivers MUST rely on the server to raise an error if
-multi-document write operations are not supported. 
+``deleteMany()`` methods as well as ``bulkWrite()`` where the requests parameter
+includes an ``UpdateMany`` or ``DeleteMany`` operation.. Drivers MUST NOT add a
```
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Extra ".".

File: source/retryable-writes/retryable-writes.rst [hunk 23]
```diff
@@ -323,79 +315,50 @@ a writable server is reselected for the retry attempt.
...
-   * events together such as retryable write operations. OPTIONAL.
-   *
-   * If set, this should be a subset of the command document containing
-   * only the sessionId and txnNum fields.
-   */
-  transactionId: Document;
-
-Drivers MUST set the ``transactionId`` field for write commands executed within
-a DriverSession where retryable writes have been enabled.
+as bulk write operations.” Each attempt of a retryable write operation SHOULD
```
Comment (ajdavis) [inline source/retryable-writes/retryable-writes.rst]: Please replace the smart quotes with straight quotes.

Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: Per our in-person chat, smart quotes are scattered throughout this repo and I was careful to only use them in prose (not code examples). Will leave as-is but a future PR can always clean them up across the repo.

File: source/retryable-writes/retryable-writes.rst (hunk 21 above)
Comment (jmikola) [reply source/retryable-writes/retryable-writes.rst]: https://github.com/mongodb/specifications/pull/174/commits/a8feb4929050b860ea21e1ba6533d6e97d39b1e1 removes "if they fail to return a response".

Depending on how drivers execute unacknowledged ops, they may or may not return a response (smallest response likely being `{ok:1}`). It's more flexible to just not refer to a possible response.
